### PR TITLE
Added test for 2pc file sink

### PR DESCRIPTION
**Pull request by olukas**

### CODE REVIEW

File: snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        if (error != null) {
+            throw new RuntimeException(error);
+        }
+        if (finished) {
+            throw new RuntimeException("[" + name + "] GeneratedFilesVerifier is not running");
+        }
+        verifier.checkStatus();
+    }
+
+    public static class Verifier {
```
Comment (gurbuzali) [inline snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java]: The logic between main thread and `Verifier` thread looks a bit complex and not optimal
- main thread reads the files and puts the lines to the `Verifier` queue
- Verifier thread then drains this queue to a `PriorityQueue`
- actual verification done on this `PriorityQueue` but if the item arrives late above process repeated

We can merge these 2 threads
- main thread reads lines and adds them to a `PriorityQueue`
- main thread then tries to verify this queue for the next step
- if everything is in order, the queue is cleared at the end of the verification
- if not, we break and main thread reads more lines and adds to the queue again

So a single thread can read and then verify them, this way we eliminate the second thread and the concurrent queue between the threads .

File: snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/FileSinkTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    @Override
+    public void init() throws Exception {
+        sleepMsBetweenItem = propertyInt("sleepMsBetweenItem", DEFAULT_SLEEP_MS_BETWEEN_ITEM);
+        snapshotIntervalMs = propertyInt("snapshotIntervalMs", DEFAULT_SNAPSHOT_INTERVAL);
+
+        stableClusterClient = Jet.newJetClient(remoteClusterClientConfig());
+    }
+
+    @Override
+    public void test() throws Throwable {
```
[the remark this answers is not in the document]
Comment (olukas) [reply snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/FileSinkTest.java]: Good idea, I'll add enhancement for that.

File: snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/FileSinkTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                    buf.add(ctx[0]++);
+                    sleepMillis(sleep);
+                })
+                .createSnapshotFn(ctx -> ctx[0])
+                .restoreSnapshotFn((ctx, state) -> ctx[0] = state.get(0))
+                .build())
+                .withTimestamps(t -> t, 0)
+                .window(WindowDefinition.tumbling(1))
+                .groupingKey(identity())
+                .aggregate(summingLong(t -> t))
```
[the remark this answers is not in the document]
Comment (olukas) [reply snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/FileSinkTest.java]: Yes, this part is there only due to item distribution to whole cluster.

[the remark this answers is not in the document]
Comment (gurbuzali) [reply snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/FileSinkTest.java]: so you don't need the timestamps actually, I use this to distribute the items generally
```
        Pipeline pipeline = Pipeline.create();

        StreamSource<Long> source = SourceBuilder
                .stream("srcForFileSink", procCtx -> new long[1])
                .<Long>fillBufferFn((ctx, buf) -> {
                    buf.add(ctx[0]++);
                    sleepMillis(sleep);
                })
                .createSnapshotFn(ctx -> ctx[0])
                .restoreSnapshotFn((ctx, state) -> ctx[0] = state.get(0))
                .build();

        pipeline.readFrom(source)
                .withoutTimestamps()
                .groupingKey(identity())
                .filterUsingService(sharedService(() -> null, ConsumerEx.noop()), (s, k, v) -> true)
                .writeTo(Sinks.filesBuilder(FILE_SINK_LOCATION_ON_MACHINE_PATH)
                              .exactlyOnce(true)
                              .build());
```

`filterUsingService` which creates a null service and does not filter anything at all

[the remark this answers is not in the document]
Comment (olukas) [reply snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/FileSinkTest.java]: Thanks Ali, I'll try to use it like you suggested.

File: snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            } catch (Throwable e) {
+                logger.severe("[" + name + "] Exception thrown during processing files.", e);
+                error = e;
+                finished = true;
+            }
+        }
+    }
+
+    private List<Long> processFiles() throws IOException {
+        try (Stream<Path> fileList = Files.list(filesinkDirectory)) {
```
Comment (gurbuzali) [inline snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java]: you don't need use `try` here, it is handled when you've finished the stream with the `collect`

**Comment by gurbuzali** [comment]
looks good overall, I've just added a very minor comment

File: snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java (hunk 4 above)
Comment (olukas) [reply snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java]: Actually I tried it without `try-with-resources` but it failed with `IOException: Too many open files` after couple of calling of `processFiles()` method. It seems that it does not close Files correctly. Using `try-with-resources` fixed this issue.

Comment (gurbuzali) [reply snapshot-file-test/src/main/java/com/hazelcast/jet/tests/snapshot/file/GeneratedFilesVerifier.java]: 👍

**Pull request merged**